Sidebar CE comments

lukasaric opened the pull request:
- This PR enables CE comments in editor/outline sidebar.
- Also implements routing & toggling CE discussions by clicking indicator button within sidebar comment

lukasaric committed:
Include CE comments in sidebar

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
CE comment badge

lukasaric committed:
Re-position badge

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Configured wrappers to work as expacted

lukasaric committed:
Set activity editor channel

lukasaric committed:
Send activity from revision component properly

lukasaric committed:
Minor badge UI tweaks :art:

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Cleanup :nail_care:

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Minor revision CE discussion btn UI tweaks

lukasaric committed:
Cleanup :lipstick:

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Minor getters & props adjustments

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Improved flaging unseen comments :surfer:

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Remove leftovers & naming polish

lukasaric committed:
Mark seen comments mutation cleanup

lukasaric committed:
Conditions polish :lipstick:

lukasaric committed:
Naming cleanup :nail_care:

lukasaric committed:
Toggle CE discussion menu within sidebar

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Improved scrolling behavior :eyes:

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Cleanup

lukasaric committed:
Handled CE dicussion toggle outside editor

lukasaric committed:
Used CE uid for scrolling/toggle purposes

lukasaric committed:
Toggle discussion cleanup :lipstick:

lukasaric committed:
Minor destructuring polish

lukasaric committed:
Removed unnecessary discussion styles on revision

lukasaric committed:
Moved CE include inside comments `list`

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
:nail_care:

lukasaric committed:
Replaced badge with btn and naming polish

lukasaric committed:
Extract discussion toggle event

lukasaric committed:
Scoped all CE comments within seen state

lukasaric committed:
Comment mutation cleanup :lipstick:

lukasaric committed:
Attached CE on cr8 & after cr8 hook

lukasaric committed:
Routing fixes & improvements

lukasaric committed:
Empty seen CE comments state

lukasaric committed:
Merge branch 'feature/ce-comments' into feature/sidebar-ce-comments

lukasaric committed:
Naming polish :scroll:

lukasaric committed:
Set last seen cleanup

lukasaric committed:
Handled after create with one reload query

lukasaric committed:
Merge branch 'develop' into feature/sidebar-ce-comments

lukasaric committed:
Prevent firing toggle if no comments

lukasaric committed:
Toggle improvements

lukasaric committed:
Routing handle enhancement :rocket:

lukasaric committed:
Improved scroll view options :bowtie:

lukasaric committed:
Disabled post btn for empty content

lukasaric committed:
Outline CE discussion routing fixes

lukasaric committed:
Merge branch 'develop' into feature/sidebar-ce-comments

lukasaric committed:
Added outline discussion seen marker :rocket:

lukasaric committed:
Improved route handling

- Added commentId query for disccussion show

lukasaric committed:
Removed unnecessart conditions :eyes:

lukasaric committed:
Added comment id as query in email CE url

lukasaric committed:
Minor timeout toggling correction

lukasaric committed:
Fixed query ids removal conditions

lukasaric committed:
Add element id query on discussion open

lukasaric committed:
Show unseen count indicator in outline

lukasaric committed:
Unseen marker/separator improvements

lukasaric committed:
Merge branch 'develop' into feature/sidebar-ce-comments

lukasaric committed:
Added missing await

lukasaric committed:
Extracted toggle route options

lukasaric committed:
Handled mark seen in one lvl event emit

lukasaric committed:
Cleanup :lipstick:

- Naming polish
- Conditions polish

lukasaric committed:
Event leftover :see_no_evil:

lukasaric committed:
Query replacement :wrench:

lukasaric committed:
Extract query comment id replacement in method

lukasaric committed:
Fixed unwanted unseen seperator behavior

lukasaric committed:
Cleanup

lukasaric committed:
Merge branch 'develop' into feature/sidebar-ce-comments

lukasaric committed:
Removed leftovers

lukasaric committed:
Prevent accessing undefined query selector

lukasaric committed:
Removed timeout for mark all as seen

lukasaric committed:
Naming :nail_care:

lukasaric committed:
Fixed trimming undefined value

underscope commented on client/components/common/tce-core/Events/DiscussionEvent.js:
```diff
@@ -1,5 +1,6 @@
   REMOVE: 'comment:remove',
-  SET_LAST_SEEN: 'comment:setLastSeen'
+  SET_LAST_SEEN: 'comment:setLastSeen',
+  TOGGLE: 'comments:toggle'
```
Try to be a bit more descriptive. It is not a general toggle.

underscope commented on client/components/common/tce-core/Discussion/Comment.vue:
```diff
@@ -1,81 +1,143 @@
-            mdi-pencil-outline
-          </v-icon>
+  <div>
+    <div v-if="showUnseenSeparator" class="unseen-separator">
```
Unseen separator should be within thread component

underscope commented on client/components/common/tce-core/Discussion/Comment.vue:
```diff
@@ -1,81 +1,143 @@
+                @click="toggleElementDiscussion"
+                color="teal"
+                text x-small>
+                Element
```
When viewed within the list, to me it looks like it relates to the element below. To me it feels like it should be part of header.

lukasaric replied:
I sent some new UI/UX proposal changes in offline discussion.

underscope commented on client/components/common/tce-core/Discussion/Comment.vue:
```diff
@@ -1,81 +1,143 @@
+            </template>
+            <span>{{ comment.createdAt | formatDate('M/D/YY h:mm A') }}</span>
+          </v-tooltip>
+          <v-tooltip left>
```
Extract (ElementLink ?) Not sure if it should toggle ? 🤔

underscope commented on client/components/common/tce-core/Discussion/Comment.vue:
```diff
@@ -1,81 +1,143 @@
+      </v-chip>
+      <div class="d-flex justify-center">
+        <transition name="slide-fade">
+          <v-btn
```
Should be part of thread not comment

underscope commented on client/components/common/tce-core/Discussion/Thread.vue:
```diff
@@ -19,7 +20,10 @@ export default {
     showAll: { type: Boolean, default: false },
-    minDisplayed: { type: Number, default: 5 }
+    minDisplayed: { type: Number, default: 5 },
+    showAllComments: { type: Boolean, default: false },
```
showAll and showAllComments ? 🤔

underscope commented on client/components/common/tce-core/Discussion/Comment.vue:
```diff
@@ -1,81 +1,143 @@
+        </transition>
+      </div>
+    </div>
+    <li class="comment">
```
List item outside of list

lukasaric committed:
Partially solved issues & basic structure refactor

lukasaric committed:
Refactor file/component discussion thread

lukasaric committed:
Thread cleanup :nail_care:

lukasaric committed:
Component naming & props cleanup

lukasaric committed:
UI/UX tweaks :art:

lukasaric committed:
Senence cased element type

lukasaric committed:
Added element selection within scroll

lukasaric committed:
Order by importance

lukasaric committed:
Used resolved CE label from registry

lukasaric committed:
cleanup

lukasaric committed:
Used Element link as comment slot

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+</script>
+
+<style lang="scss" scoped>
+.v-btn {
```
Positioning should be controlled within parent

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+
+export default {
+  name: 'element-link',
+  inject: ['$teRegistry'],
```
Label should be passed as a prop, no need to inject registry here, pass from element wrapper.

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+  props: {
+    activityId: { type: Number, required: true },
+    contentElement: { type: Object, default: () => ({}) },
+    contentElementId: { type: Number, default: null }
```
Why both?

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+      return find(this.$teRegistry._registry, { type })?.name;
+    },
+    isSameRoute: vm => vm.contentElement.uid === vm.$route.query.elementId,
+    isEditor: vm => vm.$route.name === 'editor',
```
If the parent is passing this, it is only executed once. Would like to move away in these types of components from hooking to router (should be done only within top level containers if possible)

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+    linkToElement() {
+      const { isEditor, editorRoute, isSameRoute } = this;
+      if (!isEditor) return this.$router.push(editorRoute);
+      if (isSameRoute) this.$router.replace({ query: {} });
```
Why twice (replace + push). Relict of toggle?

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,57 @@
+<template>
+  <ul class="discussion-thread">
+    <div v-for="(comment, index) in visibleItems" :key="comment.uid">
```
Wrong markup under ul

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,57 @@
+        v-bind="{ comment, user }">
+        <template v-slot:element-link>
+          <element-link
+            v-if="containAllComments && comment.contentElementId"
```
containAllComments 🤔 ? `isActivityThread` ?

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment.vue:
```diff
@@ -108,12 +117,14 @@ export default {
   &-avatar {
     width: 2.5rem;
     margin-top: 0.375rem;
+    overflow: hidden;
```
🤔

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment.vue:
```diff
@@ -108,12 +117,14 @@ export default {
   display: flex;
-  margin-bottom: 1.25rem;
+  position: relative;
+  margin-bottom: 0.75rem;
```
Control from outside 🤔 ?

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment.vue:
```diff
@@ -35,6 +39,7 @@
         <span>{{ comment.createdAt | formatDate('M/D/YY h:mm A') }}</span>
       </v-tooltip>
     </div>
+    <slot name="element-link"></slot>
```
What's the purpose / why not just include boolean prop for visibility toggle (since slot will be used only for that).

underscope commented on client/components/common/tce-core/Discussion/Thread/Separator.vue:
```diff
@@ -0,0 +1,67 @@
+<template>
+  <div class="separator-container">
+    <div v-if="showUnseenSeparator" class="unseen-separator">
```
Should be handled from outside

underscope commented on client/components/common/tce-core/Discussion/Thread/Separator.vue:
```diff
@@ -0,0 +1,67 @@
+  <div class="separator-container">
+    <div v-if="showUnseenSeparator" class="unseen-separator">
+      <v-divider class="my-3" />
+      <v-chip @click="showSeenMarker = !showSeenMarker" small outlined>
```
Rounded btn?

underscope commented on client/components/common/tce-core/Discussion/Thread/Separator.vue:
```diff
@@ -0,0 +1,67 @@
+      <div class="d-flex justify-center">
+        <transition name="slide-fade">
+          <v-btn
+            v-if="showSeenMarker"
```
Should the whole block be under v-if ?

underscope commented on client/components/common/tce-core/Discussion/Thread/Separator.vue:
```diff
@@ -0,0 +1,67 @@
+            color="teal"
+            text x-small
+            class="seen-marker">
+            <v-icon class="mr-1" x-small>mdi-check</v-icon>
```
x-small should be inherited/size set by x-small btn?

underscope commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -24,18 +24,21 @@
-      :min-displayed="commentsShownLimit"
       :show-all="showAll"
+      :min-displayed="commentsShownLimit"
+      :contain-all-comments="containAllComments"
```
Confusing name... 🤔

underscope commented on client/components/common/tce-core/Discussion/Thread/Separator.vue:
```diff
@@ -0,0 +1,67 @@
+  name: 'thread-separator',
+  props: {
+    comment: { type: Object, required: true },
+    unseenComments: { type: Array, default: () => [] },
```
No need to pass this

underscope commented on client/components/common/tce-core/Discussion/Thread/Separator.vue:
```diff
@@ -0,0 +1,67 @@
+    isAuthor: vm => vm.user.id === vm.comment.author.id,
+    showUnseenSeparator() {
+      const { comment, seenMarker, unseenComments, isAuthor } = this;
+      return seenMarker && !isAuthor && unseenComments[0]?.id === comment.id;
```
Isn't it easier for the unseenComments figure out on the upper level first unseen and pass what component needs to do 🤔 ?

underscope commented on client/store/modules/repository/comments/mutations.js:
```diff
@@ -8,12 +8,16 @@ import {
 
-const markSeenComments = (state, { activityUid, elementUid, lastCommentAt }) => {
+const markSeenComments = ({ seen }, payload) => {
+  const { activityUid, elementUid, lastCommentAt, unseenElementComments } = payload;
```
unseenElementComments 🤔 ? whats the need for this one?

underscope commented on client/store/modules/repository/comments/index.js:
```diff
@@ -6,7 +6,8 @@ const state = {
     activity: {},
-    contentElement: {}
+    contentElement: {},
+    elementComments: []
```
Why separate store?

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -93,9 +96,14 @@ export default {
+        elementUid,
+        lastCommentAt,
+        timeout,
+        unseenElementComments: unseenComments
```
Why not only timestamp of last comment?

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -73,7 +73,10 @@ export default {
       return transform(elements, (acc, it) => {
         const comments = this.getComments({ activityId, contentElementId: it.id });
-        const lastSeen = seen.contentElement[it.uid] || 0;
+        const lastSeen = {
```
Why do we need both, isn't for an element only relevant the bigger out of two?

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -139,8 +142,9 @@ export default {
         const elementId = `#element_${id}`;
-        const element = this.$refs.activityContent.querySelector(elementId);
-        element.scrollIntoView();
+        const element = this.$refs.activityContent?.querySelector(elementId);
```
When is the scrollToElement called prior to render? (`?.`)

lukasaric replied:
It's not called prior to render. It's sometimes called after the component has been unmounted (e.g, when you open editor with `elementId` query and navigate away before `isLoading` & its timeout are resolved/done. )
The only alternative way I know to this approach would be to preserve `timeoutId` and clear it on `beforeDestroy`.
@underscope thoughts? 🤔

underscope replied:
No need to change it for now

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -170,6 +174,12 @@ export default {
         items.forEach(({ elementId, ...user }) => this.selectElement(elementId, user, isSelected));
       });
+    },
+    $route(route) {
```
This is prob resulting with prev check? Need to wait for the mount.

underscope commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -57,20 +60,24 @@ export default {
     thread: vm => orderBy(vm.comments, ['createdAt'], ['asc']),
     commentsCount: vm => vm.thread.length,
     showAllToggle: vm => vm.commentsShownLimit < vm.thread.length,
+    isPostDisabled: vm => !vm.comment.content?.trim(),
```
Describe what computed is rather than what it results with for this use case; i.e. isCommentInputEmpty / bit to large, try to figure out better one...

underscope commented on client/components/repository/common/ActivityDiscussion.vue:
```diff
@@ -4,30 +4,38 @@
-      v-bind="{ comments, user, showHeading, scrollTarget: 'editor' }"
-      show-notifications />
+      @markSeen="setLastSeenComment"
+      v-bind="{ comments, user, showHeading, unseenComments, seenMarker }"
```
If both have to resolve unseen comments just pass the comments array and last seen timestamp and then resolve within the thread.

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment.vue:
```diff
@@ -58,15 +63,19 @@
     comment: { type: Object, required: true },
-    user: { type: Object, required: true }
+    user: { type: Object, required: true },
+    containAllComments: { type: Boolean, default: false }
```
Leftover?

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+      <v-btn
+        v-on="on"
+        @click="linkToElement"
+        :color="isSameRoute ? 'secondary' : 'teal'"
```
isShown ?

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+        <v-icon x-small class="ml-1">mdi-arrow-top-right-thick</v-icon>
+      </v-btn>
+    </template>
+    <span>Go to element</span>
```
View/Show element

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,58 @@
+  inject: ['$teRegistry'],
+  props: {
+    activityId: { type: Number, required: true },
+    contentElement: { type: Object, default: () => ({}) },
```
Why default to `{}`

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,57 @@
+      <thread-separator
+        @markSeen="$emit('markSeen')"
+        v-bind="{ comment, user, unseenComments, seenMarker }"
+        :is-first-comment="index === 0" />
```
Can be solved by moving later (Probably does not make sense to show unseen separator if all are unseen ?)

underscope commented on client/components/common/tce-core/Discussion/Thread/Separator.vue:
```diff
@@ -0,0 +1,67 @@
+        </transition>
+      </div>
+    </div>
+    <v-divider v-else-if="!isFirstComment" class="mt-0 mb-2 mx-1" />
```
*

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,57 @@
+    minDisplayed: { type: Number, default: 5 },
+    unseenComments: { type: Array, default: () => [] },
+    containAllComments: { type: Boolean, default: false },
+    seenMarker: { type: Boolean, default: false },
```
seen-confirmation

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -70,10 +71,12 @@ export default {
+      return comments.filter(it => {
+        const createdAt = new Date(it.createdAt).getTime();
+        return it.author.id !== user.id &&
+          createdAt > lastSeen.element &&
```
Should be only one .....

lukasaric committed:
Solved issues partially & enhancements

- Structure improvements
- Seen logic changes but still in progress
- Naming & styles polish

lukasaric committed:
Merge branch 'develop' into feature/sidebar-ce-comments

lukasaric committed:
Unseen separator cleanup

lukasaric committed:
Changed styles control within components

lukasaric committed:
Naming :eyes:

lukasaric committed:
Removed unnecessary flags & unseen polish

lukasaric committed:
Method naming & ordering

lukasaric committed:
Merge branch 'develop' into feature/sidebar-ce-comments

lukasaric committed:
Author label tweaks & stopped menu propagation

lukasaric committed:
Handled scrolling on thread intersection

lukasaric committed:
Minor scroll unseen condition coverage

lukasaric committed:
Processed unseen activity comments properly

lukasaric committed:
Unseen activity comments cleanup

lukasaric committed:
Minor condition correction

lukasaric committed:
Cleanup

lukasaric committed:
Unseen separator improvements

lukasaric committed:
UI/UX tweaks

lukasaric committed:
Integrated unseen separator in CE thread

lukasaric committed:
Ordered comments by unseen & createdAt

lukasaric committed:
Minor UI polish

lukasaric committed:
Unseen comments label cleanup

lukasaric committed:
Comment header & actions UI tweaks :art:

lukasaric committed:
Cleanup :lipstick:

lukasaric committed:
Separated comment folder in tce thread

lukasaric committed:
Leftover :see_no_evil:

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,112 @@
+        :is-focused="isEditing"
+        :show-preview="!isEditing"
+        class="content" />
+      <span v-if="isEditing" class="float-left">
```
Buttons should be consistently aligned to the right (like post comment icon; see comment on element link if that's the problem). Because of the btn padding (seen when hoverer)
![image](https://user-images.githubusercontent.com/6253820/104121499-03feb000-533f-11eb-9f88-b620a930d710.png)
buttons look misaligned

![image](https://user-images.githubusercontent.com/6253820/104121539-40321080-533f-11eb-8e3c-13d7115e16ef.png)
please fix this either by lowering the padding or changing the btn style to depressed (only the second one in this case; once aligned to the right).

underscope replied:
Consider using d-flex justify-end instead of float.

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,112 @@
+        <v-btn @click="reset" text small>Cancel</v-btn>
+        <v-btn
+          @click="save"
+          :disabled="content && !content.trim()"
```
Should we support deleting comments by editing and saving? If so, this check needs to be removed (i.e. slack allows it)

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,112 @@
+      <img :src="author.imgUrl">
+    </v-avatar>
+    <div class="comment-body pl-3">
+      <comment-header
```
Comment header should be outside of the comment body 🤔  ?
This also limits the input width and seems a bit too small
![image](https://user-images.githubusercontent.com/6253820/104122511-a7eb5a00-5345-11eb-832d-eccdd82eb64f.png)
Consider expanding it (right; empty space; can be still a bit padded on the right but not this much)

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,112 @@
+<template>
+  <div class="comment">
+    <v-avatar size="34" class="comment-avatar">
```
Avatar, user label, comment date, and actions should be part of a header.

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,112 @@
+  font-family: Roboto, Arial, sans-serif;
+
+  &-avatar {
+    width: 2.5rem;
```
Avatar size 🤔 ?

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,112 @@
+  &-avatar {
+    width: 2.5rem;
+    margin-top: 0.375rem;
+    overflow: hidden;
```
🤔 ?

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/Header.vue:
```diff
@@ -0,0 +1,51 @@
+    createdAt: { type: [Number, String], required: true }
+  },
+  computed: {
+    authorLabel: vm => vm.author.fullName || vm.author.email
```
Author should have label field attached (see user model)

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,47 @@
+<template>
```
After seeing the implications of positioning this in the footer (need to switch cancel/save on edit on left) I'd like to move this to the header after the `timeago` date and add a vertical separator or some wording between; i.e _about 2 hours ago on **HTML** element_  with HTML colored and linked using our teal accent color.

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,47 @@
+          :color="isShown ? 'teal' : 'blue-grey'"
+          text x-small>
+          {{ elementLabel }}
+          <v-icon x-small class="ml-1">mdi-arrow-top-right-thick</v-icon>
```
btn is already x-small?

lukasaric replied:
I think not. Icon size inside is decoupled from button size. Default size is medium I think. e.g (icon will not inherit `x-small` size from `btn`)

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,47 @@
+  name: 'element-link',
+  props: {
+    activityId: { type: Number, required: true },
+    contentElement: { type: Object, required: true },
```
Only uid is used

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,47 @@
+  props: {
+    activityId: { type: Number, required: true },
+    contentElement: { type: Object, required: true },
+    elementLabel: { type: String, required: true },
```
Consider renaming to label (Since component name is element link)

underscope commented on client/components/common/tce-core/Discussion/Thread/UnseenSeparator.vue:
```diff
@@ -0,0 +1,46 @@
+  <div class="unseen-separator">
+    <v-divider />
+    <v-chip
+      @click="$emit('markSeen')"
```
Just seen?

underscope commented on client/components/common/tce-core/Discussion/Thread/UnseenSeparator.vue:
```diff
@@ -0,0 +1,46 @@
+      @click:close="$emit('markSeen')"
+      close-icon="mdi-close"
+      color="teal"
+      small close outlined>
```
Group style props i.e. `outlined small close`

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+<template>
+  <ul v-intersect="onIntersect" class="discussion-thread">
+    <li
+      v-for="comment in visibleItems"
```
Consider aligning iter var name with collection name (i.e. rename comment into item or visibleItems to visibleComments)

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+        @remove="$emit('remove', comment)"
+        v-bind="{ comment, user }"
+        class="mb-3" />
+      <element-link
```
Should be slotted or part of the component

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+        v-if="isActivityThread && comment.contentElement"
+        v-bind="comment"
+        :is-editor="isEditor"
+        :element-label="elementLabel(comment)" />
```
Method name should start with a verb.

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+  },
+  data: () => ({ isVisible: false }),
+  computed: {
+    isEditor: vm => vm.$route.name === 'editor',
```
Should be decoupled from router (you can pass this as a prop). Might not be needed / see  ElementLink component.

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+    isEditor: vm => vm.$route.name === 'editor',
+    visibleItems: vm => vm.showAll ? vm.items : takeRgt(vm.items, vm.minDisplayed),
+    unseenThread: vm => orderBy(vm.unseenComments, 'createdAt', 'asc'),
+    firstUnseenComment: vm => vm.unseenThread[0]
```
Might be nicer if you extract li listing and then do a v-if for unseen block with divider (rather than doing the check for each item and having this firstUnseen logic.

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+    elementLabel({ contentElement: { type } }) {
+      return find(this.$teRegistry._registry, { type })?.name;
+    },
+    showUnseenSeparator({ id, author }) {
```
Not needed if one of the prev comments is adopted (li extraction).

underscope replied:
Instead of method you can just figure out index of firstUnseen

underscope commented on client/components/common/tce-core/ElementDiscussion.vue:
```diff
@@ -25,7 +26,8 @@
       @remove="editorBus.emit(events.REMOVE, $event)"
-      v-bind="{ comments, user }"
+      @markSeen="setLastSeen"
+      v-bind="{ comments, user, unseenComments }"
```
comments, unseenComments, user (group related)

underscope commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -7,12 +8,15 @@ export const getComments = state => params => {
-  );
+  const { items, seen } = state;
+  const lastActivitySeen = get(seen.activity, activity.uid, 0);
+  return filter(items, it => {
```
Filtering by activity first might save some additional computation and simplify the filter (early return; rather then doing it later on in if; also consider using getComments helper to make it a bit more readable rather than doing it all here)

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/Header.vue:
```diff
@@ -0,0 +1,51 @@
+      <v-tooltip right>
+        <template v-slot:activator="{ on }">
+          <span v-on="authorLabel.length > 25 ? on : null" class="author">
+            {{ authorLabel | truncate(25) }}
```
Since this will be a one-liner just use text-truncate class and the tooltip can be always shown (not a big value in doing it conditionally; filters removed in new vue version so let's avoid it when we can).

underscope commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,112 @@
+        :key="name"
+        @click="action"
+        x-small icon
+        class="mr-1">
```
ml-1 seems like a way to go since we want to use space on the right as much and need to distance actions on the left.

underscope commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -7,12 +8,15 @@ export const getComments = state => params => {
-    new Date(it.createdAt).getTime() > lastSeen
-  );
+  const { items, seen } = state;
+  const lastActivitySeen = get(seen.activity, activity.uid, 0);
```
lastActivitySeen name implies the last activity that was seen rather then specific activity last seen date

underscope commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -7,12 +8,15 @@ export const getComments = state => params => {
+    const isAuthor = it.authorId === auth.user.id;
+    const createdAt = new Date(it.createdAt).getTime();
+    if (!isCurrentActivityComment || isAuthor || lastActivitySeen >= createdAt) return;
+    if (!it.contentElement) return true;
```
Add comment why this is being done

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,47 @@
+    activityId: { type: Number, required: true },
+    contentElement: { type: Object, required: true },
+    elementLabel: { type: String, required: true },
+    isEditor: { type: Boolean, required: true }
```
Would like to remove this entirely, possibly by emitting an event rather than doing the navigation within this page (coupled at the moment).

kronicker replied:
Or you could make this component a reusable `tce-core/EditorLink` component that always navigates to editor (with optional `elementId` param)?
Elements are only shown in the editor so you don't actually need to care about the environment where the link is used and using `router-link` will take care of possible duplicated navigation errors.

underscope commented on client/components/common/tce-core/Discussion/Thread/ElementLink.vue:
```diff
@@ -0,0 +1,47 @@
+    isEditor: { type: Boolean, required: true }
+  },
+  computed: {
+    isShown: vm => vm.contentElement.uid === vm.$route.query.elementId
```
This can be probably removed (See comment above).

underscope commented on client/components/common/tce-core/Discussion/Thread/UnseenSeparator.vue:
```diff
@@ -0,0 +1,46 @@
+  ::v-deep .v-chip.v-chip--outlined.v-chip {
+    margin: -1rem 0 0.5rem 0;
+    border-radius: 1rem !important;
+    background-color: #fafafa !important;
```
setting grey lighten-5 class doesn't do the trick?

lukasaric replied:
No it does not. It will also apply color on border.

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+    scrollToFirstUnseen() {
+      this.$emit('showAll', true);
+      this.$nextTick(() => {
+        const element = this.$refs['unseen-separator'][0].$el;
```
use camelCase for refs

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+    isVisible(val) {
+      const { unseenThread, minDisplayed } = this;
+      if (!val || !unseenThread.length) return;
+      if (unseenThread.length < minDisplayed) return;
```
since repeated it can be moved to scrollToFirstUnseen (consider updating the method name aswell since it does more than that)

underscope commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,128 @@
+    display: none;
+  }
+
+  ::v-deep .element-link .v-btn {
```
Can be removed, see prev comment.

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -70,11 +71,11 @@ export default {
     activityId: vm => vm.activity.id,
     processedElements() {
-      const { elements, seen, activityId } = this;
+      const { elements, seen, activity, activityId } = this;
```
no need to pick up activityId if activity is being included ?

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -70,11 +71,11 @@ export default {
         const comments = this.getComments({ activityId, contentElementId: it.id });
-        const lastSeen = seen.contentElement[it.uid] || 0;
-        acc[it.uid] = { ...it, comments, lastSeen };
+        const allSeen = [seen.contentElement[it.uid], seen.activity[activity.uid]];
```
no need to prefix with all (seen === allSeen). In this case, it is not even allSeen so I would rather do max on this line and have the old name back. Thoughts? 🤔

lukasaric replied:
I agree. We can also rename `allSeen` to maybe `seenUids`, I think it's descriptive enough. Gather all seen `uids`, then assign `max` timestamp to lastSeen?

underscope commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -170,6 +172,12 @@ export default {
+    $route(route) {
+      const { elementId } = route.query;
+      if (!elementId) return;
+      this.selectElement(elementId);
```
We should somehow reuse this section within isLoading watcher and document when this is being used.

underscope commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -24,18 +24,21 @@
-      :min-displayed="commentsShownLimit"
       :show-all="showAll"
+      :min-displayed="commentsShownLimit"
+      :unseen-comments="unseenComments"
```
align naming (items and unseen-comments). Should be grouped as well.

underscope commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -54,17 +57,27 @@ export default {
+      const { comments, unseenComments } = this;
+      const processedThread = comments.map(comment => {
+        const unseen = unseenComments.find(it => it.id === comment.id);
+        return { ...comment, unseen: !!unseen };
```
No need to pass unseenArray if on the processed array you are passing it as attr (you can just pass items and rely on this for other usages aswell, much cleaner approach; you can also get rid of passing count and last seen dates later on)

lukasaric committed:
Partially addressed issues :rocket:

- Comment component structure refactor
- UI Tweaks
- Cleanup

lukasaric committed:
Naming improvements

lukasaric committed:
Extracted `li` as separate component

lukasaric committed:
Cleanup

lukasaric committed:
UI tweaks & added missing lodash find

lukasaric committed:
Extracted `tce-core/EditorLink`

lukasaric committed:
Get unseen activity comments getter cleanup

lukasaric committed:
Commented activity early return statement

lukasaric committed:
Getter inline note improvement

lukasaric committed:
Avatar cleanup

lukasaric committed:
Editor link color tweaks

lukasaric committed:
Unseen logic minor corrections

lukasaric committed:
Header UI tweaks

lukasaric committed:
Header cleanup :lipstick:

lukasaric committed:
Infer select & scroll element method

lukasaric committed:
Cleanup :nail_care:

lukasaric committed:
Unseen logic improvements :surfer:

lukasaric committed:
Order/group props by importance

kronicker commented on client/components/common/tce-core/Discussion/TextEditor.vue:
```diff
@@ -5,7 +5,7 @@
     <v-textarea
       v-else
-      @change="$emit('input', $event)"
+      @input="$emit('input', $event)"
```
Think about proxying events and attributes and using `v-model.trim` in the parent

kronicker commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -1,18 +1,22 @@
-  );
+export const getUnseenActivityComments = ({ seen }, getters, { auth }) => activity => {
+  const activityComments = getters.getComments({ activityId: activity.id });
+  const lastActivitySeenDate = get(seen.activity, activity.uid, 0);
```
`activitySeenAt`

kronicker commented on client/store/modules/repository/comments/getters.js:
```diff
@@ -1,18 +1,22 @@
+    if (isAuthor || lastActivitySeenDate >= createdAt) return;
+    if (!it.contentElement) return true;
+    // Return unseen activity comment if contentElement is not set
+    const lastElementSeenDate = get(seen.contentElement, it.contentElement.uid, 0);
```
`elementSeenAt`

kronicker commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -139,23 +141,31 @@ export default {
       setTimeout(() => {
         this.selectElement(elementId);
         this.scrollToElement(elementId);
+        if (!multipleSelections) return;
         this.collaboratorSelections
           .forEach(({ elementId, ...user }) => this.selectElement(elementId, user));
```
This shouldn't be in this function (semantically)

kronicker commented on client/components/editor/ActivityContent/index.vue:
```diff
@@ -70,11 +71,12 @@ export default {
         const comments = this.getComments({ activityId, contentElementId: it.id });
-        const lastSeen = seen.contentElement[it.uid] || 0;
-        acc[it.uid] = { ...it, comments, lastSeen };
+        const seenUids = [seen.contentElement[it.uid], seen.activity[activityUid]];
```
`lastSeen = max(...`

kronicker commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -54,17 +56,28 @@ export default {
+        const isUnseen = !!unseenComments.find(it => it.id === comment.id);
+        return { ...comment, isUnseen, unseenCount };
+      });
+      return orderBy(processedThread, ['isUnseen', 'createdAt'], 'asc');
```
Return object with `comments` array prop and `unseenCount`

kronicker reviewed (changes requested):
To be continued... 😉

lukasaric committed:
Naming & watchers polish

lukasaric committed:
Used trim modifier

lukasaric committed:
Separated seen/unseen lists

lukasaric committed:
Removed extra unseen arr calculation

lukasaric committed:
Cleanup

lukasaric committed:
Styles polishing

lukasaric committed:
Used partition instead of transform

lukasaric committed:
Cleanup :nail_care:

- Fixed author id issue
- Set seen on text editor focus
- Naming polish

lukasaric committed:
Added unseen-divider transition

kronicker commented on client/components/common/tce-core/Discussion/TextEditor.vue:
```diff
@@ -1,12 +1,12 @@
 <template>
```
This can be functional compoment 😉

kronicker commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,84 @@
+      this.$emit('update', comment, content);
+    },
+    remove() {
+      this.$emit('remove', this.comment);
```
I'd say this is good behavior but I would add a confirmation modal for this since it's not obvious.

kronicker commented on client/components/common/tce-core/Discussion/Thread/Comment/index.vue:
```diff
@@ -0,0 +1,84 @@
+    },
+    save() {
+      const { comment, content } = this;
+      if (!content) return this.$emit('remove', comment);
```
Either use method or inline emit in template

kronicker commented on client/components/common/tce-core/Discussion/Thread/index.vue:
```diff
@@ -0,0 +1,92 @@
+    onIntersect(_entries, _observer, isIntersected) {
+      this.isVisible = isIntersected;
+    },
+    revealUnseen(unseenCount) {
```
`count`

lukasaric committed:
Added removeconfirmation dialog & polishing

lukasaric committed:
Align text-editor margin & paddings

lukasaric committed:
Changed text-editor to functional component

lukasaric committed:
Merge branch 'develop' into feature/sidebar-ce-comments

kronicker commented on client/components/common/tce-core/Discussion/index.vue:
```diff
@@ -64,24 +66,27 @@ export default {
     },
     commentsCount: vm => vm.thread.length,
-    showAllToggle: vm => vm.commentsShownLimit < vm.thread.length,
+    showAllToggle: vm => vm.commentsShownLimit < vm.commentsCount,
```
`hasHiddenComments` or something along those lines

lukasaric committed:
Naming polish

lukasaric committed:
Added edited label left margin

kronicker reviewed (approved):
Nice job! LGTM 👌 
@underscope As far as I'm concerned this can be merged.
